File: app/src/main/java/com/chiului/android_mvvm_architecture/data/LoginRepository.java
package com.chiului.android_mvvm_architecture.data;

import com.chiului.android_mvvm_architecture.api.service.BaseApiService;
import com.chiului.android_mvvm_architecture.api.service.UserService;
import com.chiului.android_mvvm_architecture.bean.AppCacheBean;
import com.chiului.android_mvvm_architecture.bean.ApiResult;
import com.chiului.android_mvvm_architecture.utilities.AppCacheConstantsKt;
import com.google.gson.JsonObject;

import io.reactivex.rxjava3.core.Single;
import okhttp3.RequestBody;

/**
 * 用户储存库$
 *
 * @author 神经大条蕾弟
 * @date 2020/07/23 16:44
 */
public class LoginRepository {

    private LoginRepository(){}

    /**
     * Room 的数据库表对象
     */
    private static AppCacheDao mAppCacheDao;

    /**
     * Retrofit 的接口
     */
    private static UserService mUserService;

    public static LoginRepository getInstance(AppCacheDao appCacheDao, UserService userService){
        mAppCacheDao = appCacheDao;
        mUserService = userService;
        return SingletonHolder.Instance;
    }

    private static class SingletonHolder{
        private static final LoginRepository Instance = new LoginRepository();
    }

    public void saveBeforeAccount(String account){
        mAppCacheDao.insertAppCache(new AppCacheBean(AppCacheConstantsKt.ACCOUNT_BEFORE, account));
    }

    public String getBeforeAccount(){
        AppCacheBean appCache = mAppCacheDao.getAppCache(AppCacheConstantsKt.ACCOUNT_BEFORE);
        if (appCache != null) {
            return appCache.getValue();
        } else {
            return "";
        }
    }

    public void saveToken(String token){
        AppCacheBean appCacheBean = new AppCacheBean(AppCacheConstantsKt.TOKEN, token);
        mAppCacheDao.insertAppCache(appCacheBean);
    }

    public Single<ApiResult<String>> getToken(String account, String psw){
        JsonObject jsonObject = new JsonObject();
        jsonObject.addProperty("userName", account);
        jsonObject.addProperty("password", psw);
        RequestBody requestBody = BaseApiService.getJsonRequestBody(jsonObject.toString());
        Single<ApiResult<String>> single = mUserService.login(requestBody);
        return single;
    }

}

File: app/src/main/java/com/chiului/android_mvvm_architecture/data/LoginRepository.kt
package com.chiului.android_mvvm_architecture.data

import com.chiului.android_mvvm_architecture.api.service.BaseApiService
import com.chiului.android_mvvm_architecture.api.service.UserService
import com.chiului.android_mvvm_architecture.bean.ApiResult
import com.chiului.android_mvvm_architecture.bean.AppCacheBean
import com.chiului.android_mvvm_architecture.utilities.ACCOUNT_BEFORE
import com.chiului.android_mvvm_architecture.utilities.TOKEN
import com.google.gson.JsonObject
import io.reactivex.rxjava3.core.Single

/**
 * 用户存储库$
 * @author    神经大条蕾弟
 * @date      2021/01/29 16:58
 */
class LoginRepository private constructor(
        private val appCacheDao: AppCacheDao,
        private val userService: UserService
) {

    companion object {
        // For Singleton instantiation
        @Volatile
        private var instance: LoginRepository? = null

        fun getInstance(appCacheDao: AppCacheDao, userService: UserService) =
                instance ?: synchronized(this) {
                    instance ?: LoginRepository(appCacheDao, userService).also { instance = it }
                }
    }

    /**
     * 保存上次登录账号
     * @param account String 登录成功的账号
     */
    fun saveBeforeAccount(account: String) {
        appCacheDao.insertAppCache(AppCacheBean(ACCOUNT_BEFORE, account))
    }

    /**
     * 获取上次登录的账号
     * @return String 上次登录的账号
     */
    fun getBeforeAccount(): String {
        val appCache: AppCacheBean? = appCacheDao.getAppCache(ACCOUNT_BEFORE)
        return appCache?.value ?: ""
    }

    /**
     * 保存登录成功的 token
     * @param token String token
     */
    fun saveToken(token: String) {
        val appCacheBean = AppCacheBean(TOKEN, token)
        appCacheDao.insertAppCache(appCacheBean)
    }

    /**
     * 从后台获取 token
     * @param account String?
     * @param psw String?
     * @return Single<ApiResult<String>>
     */
    fun getRemoteToken(account: String?, psw: String?): Single<ApiResult<String>> {
        val jsonObject = JsonObject()
        jsonObject.addProperty("userName", account)
        jsonObject.addProperty("password", psw)
        val requestBody = BaseApiService.getJsonRequestBody(jsonObject.toString())
        return userService.login(requestBody)
    }

    /**
     * 获取本地数据库 token
     */
    fun getLocalToken(): AppCacheBean? {
        return appCacheDao.getAppCache(TOKEN)
    }

    /**
     * 删除本地 token
     */
    fun deleteToken() {
        val tokenCache: AppCacheBean? = getLocalToken()
        if (tokenCache != null) {
            appCacheDao.deleteAppCache(tokenCache)
        }
    }

}